torch/_inductor/codegen/cpp.py

```diff
@@ -2005,115 +2068,96 @@ def codegen_loops_impl(self, loop_nest, code, worksharing):
                 if isinstance(kernel, OuterLoopFusedKernel):
-                    for loop in kernel.inner:
-                        if loop.inner:
-                            gen_loops(loop.inner, loop.is_reduction)
-                        else:
-                            with contextlib.ExitStack() as stack:
-                                # If there is any kernel existing at the final outer loop fusion level,
-                                # the kernel code should be placed within its respective indent to prevent
-                                # the duplication of variable definitions.
-                                stack.enter_context(code.indent())
-                                gen_loop_kernel(loop)
+                    for _loop_nest in kernel.inner:
+                        gen_loop_nest(_loop_nest)
                 else:
+                    if _loop_nest.loops is not None and is_parallel_reduction():
+                        assert isinstance(kernel, CppKernelProxy)
+                        kernel.update_stores_with_parallel_reduction()
                     with contextlib.ExitStack() as stack:
-                        assert kernel
-                        if hasattr(kernel, "codegen_inner_loops"):
-                            code.splice(kernel.preloads)
-                            kernel.codegen_inner_loops(code)
-                            stack.enter_context(code.indent())
-                        code.splice(kernel.loads)
-                        code.splice(kernel.compute)
-                        code.splice(kernel.stores)
-                    if hasattr(kernel, "codegen_inner_loops"):
-                        code.splice(kernel.poststores)
-
-            def get_reduction_code_buffer(loops, buffer="prefix"):
-                assert buffer in ("prefix", "suffix", "local")
-                for loop in loops:
-                    for kernel in loop.get_kernels():
-                        if buffer == "local":
-                            return (
-                                kernel.local_reduction_init,
-                                kernel.local_reduction_stores,
-                            )
-                        elif buffer == "suffix":
-                            suffix = kernel.reduction_suffix
-                            if loop.parallel:
-                                suffix = kernel.parallel_reduction_suffix + suffix
-                            return suffix
-                        else:
-                            prefix = kernel.reduction_prefix
-                            if loop.parallel:
-                                prefix = prefix + kernel.parallel_reduction_prefix
-                            else:
-                                prefix = prefix + kernel.non_parallel_reduction_prefix
-                            return prefix
+                        stack.enter_context(code.indent())
+                        kernel.gen_body(code)
```

Comment:
Assert the kernel must be the instance of `CppKernelProxy`?

Reply:
Done.